### PR TITLE
Implement token analyzer

### PR DESCRIPTION
**Pull request by zizou0x**

### CODE REVIEW

File: token-analyzer/src/http_client.rs [hunk 1]
```diff
@@ -0,0 +1,120 @@
+use anyhow::{anyhow, Result};
+use reqwest::{Client, ClientBuilder, Response};
+use std::{
+    fmt::{self, Display, Formatter},
+    time::Duration,
+};
+
+const USER_AGENT: &str = "cowprotocol-services/2.0.0";
```
Comment (kayibal) [inline token-analyzer/src/http_client.rs]: Let's change this I see this causing confusion later down the line.

File: substreams/ethereum-ambient/substreams.yaml [hunk 2]
```diff
@@ -19,34 +19,33 @@ binaries:
 modules:
   - name: map_pool_changes
     kind: map
-    initialBlock: 17361664
+    initialBlock: 17381533
```
Comment (kayibal) [inline substreams/ethereum-ambient/substreams.yaml]: Why did you change these?

File: token-analyzer/src/main.rs [hunk 3]
```diff
@@ -33,15 +19,25 @@ async fn main() -> Result<(), ()> {
         "transport".to_owned(),
     ));
     let w3 = Web3::new(transport);
-    let tf = TokenFinder {};
+    let tf = TokenFinder::new(HashMap::from([(
```
Comment (kayibal) [inline token-analyzer/src/main.rs]: Let's remove main from token-analyzer or better let's put it under examples?

File: token-analyzer/src/trace_call.rs [hunk 4]
```diff
@@ -28,16 +28,20 @@ pub struct TraceCallDetector {
...
 }
 
 #[async_trait::async_trait]
 impl BadTokenDetecting for TraceCallDetector {
-    async fn detect(&self, token: H160) -> Result<(TokenQuality, Option<U256>, Option<U256>)> {
-        let quality = self.detect_impl(token).await?;
+    async fn detect(
+        &self,
+        token: H160,
+        block: BlockNumber,
```
Comment (kayibal) [inline token-analyzer/src/trace_call.rs:42]: Can we maybe use our types in public interfaces and cast them to the required types in the implementations?

File: tycho-indexer/src/extractor/evm/native.rs [hunk 5]
```diff
@@ -254,9 +256,27 @@ where
...
         if !new_tokens_addresses.is_empty() {
+            let balance_map: HashMap<H160, (H160, U256)> = balance_changes
+                .iter()
+                .filter_map(|bc| match H160::from_str(bc.component_id.as_str()) {
+                    Ok(address) => Some((
+                        bc.token.clone().into(),
+                        (address, U256::from_big_endian(&bc.new_balance.clone().to_vec())),
+                    )),
+                    Err(e) => {
+                        warn!("Error parsing component_id to H160: {}", e);
```
Comment (kayibal) [inline tycho-indexer/src/extractor/evm/native.rs:267]: This should be debug or maybe even trace. If we fail to analyze a token e.g. because after this no balances are left, in that case we can issue a warn.

File: tycho-indexer/src/extractor/evm/token_pre_processor.rs [hunk 6]
```diff
@@ -1,40 +1,64 @@
...
 }
 #[cfg_attr(test, mockall::automock)]
 #[async_trait]
 pub trait TokenPreProcessorTrait: Send + Sync {
-    async fn get_tokens(&self, addresses: Vec<H160>) -> Vec<ERC20Token>;
+    async fn get_tokens(
+        &self,
+        addresses: Vec<H160>,
+        token_finder: Arc<dyn TokenOwnerFinding>,
+        block: BlockNumber,
```
Comment (kayibal) [inline tycho-indexer/src/extractor/evm/token_pre_processor.rs:35]: Can we use internal type here as well?

File: substreams/ethereum-ambient/substreams.yaml (hunk 2 above)
Comment (zizou0x) [reply substreams/ethereum-ambient/substreams.yaml]: mb

File: token-analyzer/src/trace_call.rs (hunk 4 above)
Comment (zizou0x) [reply token-analyzer/src/trace_call.rs:42]: I also don't like it but let's keep it for cd week

File: tycho-indexer/src/extractor/evm/native.rs (hunk 5 above)
Comment (zizou0x) [reply tycho-indexer/src/extractor/evm/native.rs:267]: I would prefer to keep it a warning now, as we really don't expect this to happen, and change it to debug if it's becomes too much (like if we add a protocol that breaks this basically)
